File: app/src/main/java/org/codepond/imdemo/chat/ChatPresenter.java
package org.codepond.imdemo.chat;

import android.support.annotation.Nullable;

import org.codepond.imdemo.ChatMessage;
import org.codepond.imdemo.service.chat.MessagingService;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Named;

class ChatPresenter implements ChatContracts.Presenter, MessagingService.OnMessageReceivedListener {
    private final String mUserJid;
    private final ChatContracts.View mView;
    private final String mParticipantJid;
    private final List<ChatMessage> mMessages;
    @Nullable private MessagingService mMessagingService;

    @Inject
    ChatPresenter(ChatContracts.View view,
                  @Named("userJid") String userJid,
                  @Named("participantJid") String participantJid) {
        mView = view;
        mUserJid = userJid;
        mParticipantJid = participantJid;
        mMessages = new ArrayList<>();
    }

    @Override
    public void loadMessages() {
        // TODO: Load message history from DB
        mView.showMessages(mMessages);
    }

    @Override
    public void sendMessage(String message) {
        if (message != null && message.length() > 0) {
            ChatMessage chatMessage = new ChatMessage(mUserJid, mParticipantJid, message, false, System.currentTimeMillis());
            mMessages.add(chatMessage);
            mView.notifyNewMessageAdded();
            mView.cleanUserInput();
            if (mMessagingService != null) {
                mMessagingService.sendMessage(chatMessage);
            }
            // TODO: store message in DB
        }
    }

    @Override
    public void onMessageReceived(ChatMessage chatMessage) {
        // TODO: store message in DB
        mMessages.add(chatMessage);
        mView.notifyNewMessageAdded();
    }

    List<ChatMessage> getMessages() {
        return mMessages;
    }

    @Override
    public void start(MessagingService messagingService) {
        if (messagingService == null) {
            throw new NullPointerException("messagingService must not be null!");
        }
        mMessagingService = messagingService;
        mMessagingService.setCurrentParticipant(mParticipantJid);
        mMessagingService.setOnMessageReceivedListener(this);
    }

    @Override
    public void stop() {
        if (mMessagingService != null) {
            mMessagingService.setOnMessageReceivedListener(null);
        }
    }
}

File: app/src/main/java/org/codepond/imdemo/chat/ChatPresenter.kt
package org.codepond.imdemo.chat

import org.codepond.imdemo.ChatMessage
import org.codepond.imdemo.service.chat.MessagingService

import java.util.ArrayList

import javax.inject.Inject
import javax.inject.Named

class ChatPresenter
@Inject
constructor(private val view: ChatContracts.View,
            @Named("userJid") private val userJid: String,
            @Named("participantJid") private val participantJid: String) : ChatContracts.Presenter, MessagingService.OnMessageReceivedListener {

    val messages: MutableList<ChatMessage> = ArrayList()

    private var messagingService: MessagingService? = null

    override fun loadMessages() {
        // TODO: Load message history from DB
        view.showMessages(messages)
    }

    override fun sendMessage(message: String) {
        if (message.isNotEmpty()) {
            val chatMessage = ChatMessage(userJid, participantJid, message, false, System.currentTimeMillis())
            messages.add(chatMessage)
            view.notifyNewMessageAdded()
            view.cleanUserInput()
            messagingService?.sendMessage(chatMessage)
            // TODO: store message in DB
        }
    }

    override fun onMessageReceived(chatMessage: ChatMessage) {
        // TODO: store message in DB
        messages.add(chatMessage)
        view.notifyNewMessageAdded()
    }

    override fun start(messagingService: MessagingService) {
        messagingService.setCurrentParticipant(participantJid)
        messagingService.setOnMessageReceivedListener(this)
        this.messagingService = messagingService
    }

    override fun stop() {
        messagingService?.setOnMessageReceivedListener(null)
    }
}
